marklogic-data-hub-central/src/test/java/com/marklogic/hub/central/controllers/steps/CustomStepControllerTest.java

```diff
@@ -29,41 +29,137 @@
     		+ " \"sourceQuery\": \"cts.collectionQuery('default')\" , "
     		+ " \"targetEntityType\": \"http://example.org/Customer-0.0.1/Customer\" , "
     		+ " \"stepDefinitionName\": \"custom\" , "
+            + " \"stepDefinitionType\": \"custom\" , "
+            + " \"stepId\": \"myCustomStep-custom\" , "
     		+ " \"batchSize\": \"50\" , "
     		+ " \"provenanceGranularityLevel\": \"coarse\" "
     		+ "}";
     private final static ObjectMapper MAPPER = new ObjectMapper();
-
-    public static class CustomStep {
-        public String name;
-        public String description;
-        public String selectedSource;
-        public String sourceQuery;
-        public String targetEntityType;
-        public String stepDefinitionName;
-    }
-
-    public static CustomStep newDefaultCustomStep(String name) {
-        CustomStep step = new CustomStep();
-        step.name = name;
-        step.description = "optional";
-        step.selectedSource = "collection";
-        step.sourceQuery = "cts.collectionQuery('test')";
-        step.targetEntityType = "http://example.org/Customer-0.0.1/Customer";
-        step.stepDefinitionName = "custom";
-        return step;
+    
+    @Test
+    void userCanWrite() throws Exception {
+    	AtomicReference<JsonNode> stepData = new AtomicReference<>();

```

Comment:
There is no need to use "AtomicReference". I have rewritten the entire test.Please make the necessary changes.

```
    @Test
    void permittedWriteUser() throws Exception {
        installOnlyReferenceModelEntities();
        loginAsTestUserWithRoles("hub-central-custom-writer");

        /* create new custom step (not supported by CustomStepController) */
        JsonNode newCustomStepNode = MAPPER.readTree("{"
        + " \"name\": \"myCustomStep\" , "
            + " \"description\": \"\" , "
            + " \"selectedSource\": \"\" , "
            + " \"sourceQuery\": \"cts.collectionQuery('default')\" , "
            + " \"targetEntityType\": \"http://example.org/Customer-0.0.1/Customer\" , "
            + " \"stepDefinitionName\": \"custom\" , "
            + " \"stepDefinitionType\": \"custom\" , "
            + " \"stepId\": \"myCustomStep-custom\" , "
            + " \"batchSize\": \"50\" , "
            + " \"provenanceGranularityLevel\": \"coarse\" , "
            + " \"prop\": \"value\" , " // additional settings
            + " \"prop2\": \"value2\" "
            + "}");
        StepService.on(getHubClient().getStagingClient()).saveStep("custom", newCustomStepNode, false, true);

        getJson(PATH + "/myCustomStep").andExpect(status().isOk())
            .andDo(result -> updateCustomStep(parseJsonResponse(result)));
        
        getJson(PATH + "/myCustomStep").andExpect(status().isOk())
            .andDo(result -> {
                JsonNode updatedCustomStep =  parseJsonResponse(result) ;
                assertTrue(updatedCustomStep.get("name").asText().equals("myCustomStep"));
                assertTrue(updatedCustomStep.get("sourceQuery").asText().equals("no_query"));
                assertNotNull(updatedCustomStep.get("additionalSettings"));
                assertTrue(updatedCustomStep.get("additionalSettings").get("prop").asText().equals("foo"));
                assertNull(updatedCustomStep.get("additionalSettings").get("prop2"));
            });
    }

    private void updateCustomStep(JsonNode customStepNode) throws Exception {
        ObjectNode updatedCustomStepNode = (ObjectNode) objectMapper.readTree(customStepNode.asText());
        ObjectNode additionalSettings = (ObjectNode) customStepNode.get("additionalSettings");

        additionalSettings.put("prop", "foo");
        additionalSettings.remove("prop2");

        updatedCustomStepNode.set("additionalSettings", additionalSettings);
        updatedCustomStepNode.put("sourceQuery", "no_query");
        putJson(PATH + "/myCustomStep", updatedCustomStepNode).andExpect(status().isOk());
    }
```